sql: audit all usages of Query to use iterator pattern

yuzefovich opened the pull request:
Similarly to the previous commit (dbc8676679d21d3f61568d687e74b7c83329adb2), here we audit all usages of `Query`
method of the internal executor to take advantage of the iterator API
wherever possible (or switching to `Exec` or `QueryRow`).
`QueryBuffered` has been added to the interface too.

The only place where it would be beneficial to use the iterator pattern
but it is not done currently is for `SHOW STATISTICS` statement - in
there, we have a panic-catcher which works only on the assumption of not
updating any of the shared state (which the iterator API contradicts).
Refactoring that part is left as a TODO.

Fixes: #48595.

Release justification: low-risk update to existing functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59339)
<!-- Reviewable:end -->

yuzefovich committed:
sql: audit all usages of Query to use iterator pattern

Similarly to the previous commit, here we audit all usages of `Query`
method of the internal executor to take advantage of the iterator API
wherever possible (or switching to `Exec` or `QueryRow`).
`QueryBuffered` has been added to the interface too.

The only place where it would be beneficial to use the iterator pattern
but it is not done currently is for `SHOW STATISTICS` statement - in
there, we have a panic-catcher which works only on the assumption of not
updating any of the shared state (which the iterator API contradicts).
Refactoring that part is left as a TODO.

Release justification: low-risk update to existing functionality.

Release note: None

yuzefovich commented:
@irfansharif @otan @rafiss github suggested you as reviewers. Could one of you take a look at this please?

rafiss commented:
CCing @RichardJCai too as it may touch his idea in https://github.com/cockroachdb/cockroach/issues/60507

irfansharif commented:
(Looking at the code areas, I'm definitely not familiar with any of it - so I'll defer elsewhere. Sorry!)

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 22 of 22 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59339#-:-MUo1bmj-8PcAlaXx5oJ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan, @rafiss, @RichardJCai, and @yuzefovich)

___
*[pkg/ccl/backupccl/backup_planning.go, line 961 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59339#-MUo0h9923GiNfUg22rb:-MUo0h9923GiNfUg22rc:bk6rwf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/10ede958bdee90c8dace86e64a71b5d86259e014/pkg/ccl/backupccl/backup_planning.go#L961)):*
> ```Go
> 			// Include all tenants.
> 			// TODO(tbg): make conditional on cluster setting.
> 			tenantRows, err = p.ExecCfg().InternalExecutor.QueryBuffered(
> ```

This could be problematic with many tenants, right?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59339#-:-MUo3_8G9Uj8TMAo4amj:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @otan, @rafiss, and @RichardJCai)

___
*[pkg/ccl/backupccl/backup_planning.go, line 961 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59339#-MUo0h9923GiNfUg22rb:-MUo3N7p9Ruy_ow3ucC9:bdiios0) ([raw file](https://github.com/cockroachdb/cockroach/blob/10ede958bdee90c8dace86e64a71b5d86259e014/pkg/ccl/backupccl/backup_planning.go#L961)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This could be problematic with many tenants, right?
</blockquote></details>

Yeah, possibly, but it didn't seem trivial to refactor, so I'll leave it to Bulk IO friends. cc @dt @adityamaru 


<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
